src/atlas_protocol_scripts/pp/_check_metadata.py

```diff
@@ -0,0 +1,112 @@
+
+    # Check if keys are present as columns and verify values if present (except keys_to_ignore)
+    for key, value in ref_meta_dict.items():
+        if key in keys_to_ignore:
+            continue  # Skip keys to ignore
+
+        if key not in adata.obs.columns:
+            raise ValueError(f"Missing columns in adata.obs: {key}")
+
+        # Verify type of corresponding column
+        column_type = adata.obs[key].dtype
+        expected_type = value.get("type", None)
+
+        if expected_type is not None and column_type != expected_type:
+            offending_value = adata.obs[key][adata.obs[key].apply(lambda x: type(x) != expected_type)].iloc[0]
+            raise ValueError(
+                f"Unexpected data type found in column '{key}'. Expected '{expected_type}', but found '{offending_value}'."
+            )
+
+        # Verify values in corresponding column
+        allowed_values = value.get("values", None)
+        column_values = adata.obs[key].unique()
+
+        if "min" in value and "max" in value:
+            min_value = value["min"]
+            max_value = value["max"]
+            invalid_values = [val for val in column_values if not (min_value <= val <= max_value)]
+            if invalid_values:
+                raise ValueError(f"Invalid values found in column '{key}': {invalid_values}")
+        elif allowed_values is not None:
+            invalid_values = [val for val in column_values if val not in allowed_values]
+            if invalid_values:
+                raise ValueError(f"Invalid values found in column '{key}': {invalid_values}")
+
+        # Verify values in corresponding column
+        allowed_values = value.get("values", None)
+        column_values = adata.obs[key].unique()
+
+        if "min" in value and "max" in value:
+            min_value = value["min"]
+            max_value = value["max"]
+            invalid_values = [val for val in column_values if not (min_value <= val <= max_value)]
+            if invalid_values:
+                raise ValueError(f"Invalid values found in column '{key}': {invalid_values}")
+        elif allowed_values is not None:
+            invalid_values = [val for val in column_values if val not in allowed_values]
+            if invalid_values:
+                raise ValueError(f"Invalid values found in column '{key}': {invalid_values}")
+
+
+def search_dict(my_dict: dict, columns: List[str], search: Optional[List[str]] = None) -> dict:
+    """Searches a nested dictionary for specified keys in each of the columns."""
+    values = {}
+    for column in columns:
+        if column in my_dict:
+            column_values = {}
+            for key, value in my_dict[column].items():
+                if not search or key in search:
+                    if key in ["values", "description", "type"]:
+                        column_values[key] = value
+            values[column] = column_values
+    return values
```

Comment:
I don't quite get what this does and why this is useful.

Reply:
When harmonizing metadata on a specific `adata` I use this to quickly remind me what are approved input values.
For example, if I don't remember what the allowed values are for column `"condition"` I can quickly have a look with 
`aps.pp.search_dict(ref_meta_dict, ["condition"])`
This will give:
```python
{'condition': {'values': ['LUAD', 'LSCC', 'NSCLC'],
  'description': 'Lung adenocarcinoma (LUAD) and lung squamous cell carcinoma (LSCC) are the most common subtypes of non-small-cell lung cancer (NSCLC)'}}
```

For me this is a convenient way to quickly convert metadata into the appropriate type without needing to look through the whole dict, yaml, or needing to write a loop every time.
But we can remove it for the atlas_protocol.

Reply:
I see how this can be useful, but the examples in the notebook can be replaced by a simple
```
ref_meta_dict["condition"]
```

So I'd go with that to keep things simple.

Reply:
just print both?
```
ref_meta_dict["condition"]
ref_meta_dict["dataset"]
```

Or if you want to go fancy
```
{**ref_meta_dict["condition"], **ref_meta_dict["dataset"]}
```

Reply:
fair enough ;)